File: Aether Mobile SDK/Ecommerce.swift
// =============================================================================
// AETHER SDK — E-COMMERCE MODULE (iOS)
// Revenue tracking: product views, cart, checkout funnel, purchases
// =============================================================================

import Foundation

// MARK: - Types

public struct AetherProduct: Codable {
    public let id: String
    public let name: String
    public var category: String?
    public var brand: String?
    public var variant: String?
    public var price: Double
    public var quantity: Int?
    public var currency: String?
    public var coupon: String?

    public init(
        id: String,
        name: String,
        category: String? = nil,
        brand: String? = nil,
        variant: String? = nil,
        price: Double = 0,
        quantity: Int? = nil,
        currency: String? = nil,
        coupon: String? = nil
    ) {
        self.id = id
        self.name = name
        self.category = category
        self.brand = brand
        self.variant = variant
        self.price = price
        self.quantity = quantity
        self.currency = currency
        self.coupon = coupon
    }

    func toProperties() -> [String: AnyCodable] {
        var props: [String: AnyCodable] = [
            "product_id": AnyCodable(id),
            "name": AnyCodable(name),
            "price": AnyCodable(price)
        ]
        if let category = category { props["category"] = AnyCodable(category) }
        if let brand = brand { props["brand"] = AnyCodable(brand) }
        if let variant = variant { props["variant"] = AnyCodable(variant) }
        if let quantity = quantity { props["quantity"] = AnyCodable(quantity) }
        if let currency = currency { props["currency"] = AnyCodable(currency) }
        if let coupon = coupon { props["coupon"] = AnyCodable(coupon) }
        return props
    }
}

public struct AetherCartItem: Codable {
    public let product: AetherProduct
    public let quantity: Int

    public init(product: AetherProduct, quantity: Int) {
        self.product = product
        self.quantity = max(1, quantity)
    }

    func toProperties() -> [String: AnyCodable] {
        var props = product.toProperties()
        props["quantity"] = AnyCodable(quantity)
        return props
    }
}

public struct AetherOrder: Codable {
    public let orderId: String
    public let revenue: Double
    public var tax: Double?
    public var shipping: Double?
    public var discount: Double?
    public var coupon: String?
    public var currency: String?
    public var products: [AetherProduct]

    public init(
        orderId: String,
        revenue: Double,
        tax: Double? = nil,
        shipping: Double? = nil,
        discount: Double? = nil,
        coupon: String? = nil,
        currency: String? = nil,
        products: [AetherProduct] = []
    ) {
        self.orderId = orderId
        self.revenue = revenue
        self.tax = tax
        self.shipping = shipping
        self.discount = discount
        self.coupon = coupon
        self.currency = currency
        self.products = products
    }

    func toProperties() -> [String: AnyCodable] {
        var props: [String: AnyCodable] = [
            "order_id": AnyCodable(orderId),
            "revenue": AnyCodable(revenue)
        ]
        if let tax = tax { props["tax"] = AnyCodable(tax) }
        if let shipping = shipping { props["shipping"] = AnyCodable(shipping) }
        if let discount = discount { props["discount"] = AnyCodable(discount) }
        if let coupon = coupon { props["coupon"] = AnyCodable(coupon) }
        if let currency = currency { props["currency"] = AnyCodable(currency) }
        if !products.isEmpty {
            props["products"] = AnyCodable(products.map { $0.toProperties().mapValues { $0.value } })
        }
        return props
    }
}

// MARK: - AetherEcommerce

public final class AetherEcommerce {
    public static let shared = AetherEcommerce()

    private let serialQueue = DispatchQueue(label: "com.aether.sdk.ecommerce")
    private let defaults = UserDefaults(suiteName: "com.aether.sdk")!
    private let cartKey = "aether_cart"

    /// In-memory cart keyed by product ID.
    private var cart: [String: AetherCartItem] = [:]

    private init() {
        loadCart()
    }

    // MARK: - Product Events

    /// Track when a user views a product.
    public func productViewed(_ product: AetherProduct) {
        Aether.shared.track("product_viewed", properties: product.toProperties())
    }

    /// Track when a user views a product list / collection.
    public func productListViewed(listId: String, products: [AetherProduct]) {
        var props: [String: AnyCodable] = [
            "list_id": AnyCodable(listId),
            "products": AnyCodable(products.map { $0.toProperties().mapValues { $0.value } })
        ]
        props["product_count"] = AnyCodable(products.count)
        Aether.shared.track("product_list_viewed", properties: props)
    }

    // MARK: - Cart Management

    /// Add a product to the cart (or update quantity if already present).
    public func addToCart(_ product: AetherProduct, quantity: Int = 1) {
        serialQueue.async { [weak self] in
            guard let self = self else { return }
            let item = AetherCartItem(product: product, quantity: quantity)
            if let existing = self.cart[product.id] {
                let merged = AetherCartItem(
                    product: product,
                    quantity: existing.quantity + quantity
                )
                self.cart[product.id] = merged
            } else {
                self.cart[product.id] = item
            }
            self.persistCart()
        }

        var props = product.toProperties()
        props["quantity"] = AnyCodable(quantity)
        Aether.shared.track("add_to_cart", properties: props)
    }

    /// Remove a product from the cart by its ID.
    public func removeFromCart(productId: String) {
        var removedProps: [String: AnyCodable] = ["product_id": AnyCodable(productId)]
        serialQueue.sync {
            if let item = cart[productId] {
                removedProps = item.toProperties()
                cart.removeValue(forKey: productId)
                persistCart()
            }
        }
        Aether.shared.track("remove_from_cart", properties: removedProps)
    }

    /// Track a cart-viewed event with all current items.
    public func cartViewed() {
        let snapshot = serialQueue.sync { cartSnapshot() }
        Aether.shared.track("cart_viewed", properties: snapshot)
    }

    /// Get all items currently in the cart.
    public func getCart() -> [AetherCartItem] {
        return serialQueue.sync { Array(cart.values) }
    }

    /// Calculate total cart value (price * quantity for each item).
    public func getCartValue() -> Double {
        return serialQueue.sync {
            cart.values.reduce(0.0) { $0 + ($1.product.price * Double($1.quantity)) }
        }
    }

    /// Remove all items from the cart.
    public func clearCart() {
        serialQueue.async { [weak self] in
            self?.cart.removeAll()
            self?.persistCart()
        }
    }

    // MARK: - Checkout Funnel

    /// Track the start of a checkout.
    public func checkoutStarted(properties: [String: AnyCodable]? = nil) {
        var props = serialQueue.sync { cartSnapshot() }
        if let extra = properties {
            props.merge(extra) { _, new in new }
        }
        Aether.shared.track("checkout_started", properties: props)
    }

    /// Track individual checkout steps (e.g., shipping, payment, review).
    public func checkoutStepCompleted(step: Int, properties: [String: AnyCodable]? = nil) {
        var props: [String: AnyCodable] = ["step": AnyCodable(step)]
        if let extra = properties {
            props.merge(extra) { _, new in new }
        }
        Aether.shared.track("checkout_step_completed", properties: props)
    }

    // MARK: - Order Events

    /// Track a completed purchase.
    public func orderCompleted(_ order: AetherOrder) {
        Aether.shared.track("order_completed", properties: order.toProperties())

        // Clear cart after successful order
        clearCart()
    }

    /// Track a full or partial refund.
    public func orderRefunded(orderId: String, products: [AetherProduct]? = nil) {
        var props: [String: AnyCodable] = ["order_id": AnyCodable(orderId)]
        if let products = products, !products.isEmpty {
            props["products"] = AnyCodable(products.map { $0.toProperties().mapValues { $0.value } })
            props["full_refund"] = AnyCodable(false)
        } else {
            props["full_refund"] = AnyCodable(true)
        }
        Aether.shared.track("order_refunded", properties: props)
    }

    // MARK: - Coupons

    /// Track when a coupon is applied.
    public func couponApplied(coupon: String, discount: Double? = nil) {
        var props: [String: AnyCodable] = ["coupon": AnyCodable(coupon)]
        if let discount = discount {
            props["discount"] = AnyCodable(discount)
        }
        Aether.shared.track("coupon_applied", properties: props)
    }

    // MARK: - Persistence

    private func cartSnapshot() -> [String: AnyCodable] {
        let items = cart.values.map { $0.toProperties().mapValues { $0.value } }
        return [
            "products": AnyCodable(items),
            "cart_value": AnyCodable(cart.values.reduce(0.0) { $0 + ($1.product.price * Double($1.quantity)) }),
            "item_count": AnyCodable(cart.count)
        ]
    }

    private func persistCart() {
        guard let data = try? JSONEncoder().encode(Array(cart.values)) else { return }
        defaults.set(data, forKey: cartKey)
    }

    private func loadCart() {
        guard let data = defaults.data(forKey: cartKey),
              let items = try? JSONDecoder().decode([AetherCartItem].self, from: data) else { return }
        for item in items {
            cart[item.product.id] = item
        }
    }
}

File: Aether Mobile SDK/Ecommerce.kt
// =============================================================================
// AETHER SDK — E-COMMERCE MODULE (Android)
// Revenue tracking: product views, cart, checkout funnel, purchases
// =============================================================================

package com.aether.sdk

import android.content.Context
import android.content.SharedPreferences
import org.json.JSONArray
import org.json.JSONObject
import java.util.concurrent.ConcurrentHashMap

// =============================================================================
// TYPES
// =============================================================================

data class AetherProduct(
    val id: String,
    val name: String,
    val category: String? = null,
    val brand: String? = null,
    val variant: String? = null,
    val price: Double = 0.0,
    val quantity: Int? = null,
    val currency: String? = null,
    val coupon: String? = null
) {
    fun toMap(): Map<String, Any?> {
        val map = mutableMapOf<String, Any?>(
            "product_id" to id,
            "name" to name,
            "price" to price
        )
        category?.let { map["category"] = it }
        brand?.let { map["brand"] = it }
        variant?.let { map["variant"] = it }
        quantity?.let { map["quantity"] = it }
        currency?.let { map["currency"] = it }
        coupon?.let { map["coupon"] = it }
        return map
    }

    fun toJSONObject(): JSONObject = JSONObject(toMap().filterValues { it != null })
}

data class AetherCartItem(
    val product: AetherProduct,
    val quantity: Int
) {
    fun toMap(): Map<String, Any?> {
        val map = product.toMap().toMutableMap()
        map["quantity"] = quantity
        return map
    }

    fun toJSONObject(): JSONObject = JSONObject(toMap().filterValues { it != null })
}

data class AetherOrder(
    val orderId: String,
    val revenue: Double,
    val tax: Double? = null,
    val shipping: Double? = null,
    val discount: Double? = null,
    val coupon: String? = null,
    val currency: String? = null,
    val products: List<AetherProduct> = emptyList()
) {
    fun toMap(): Map<String, Any?> {
        val map = mutableMapOf<String, Any?>(
            "order_id" to orderId,
            "revenue" to revenue
        )
        tax?.let { map["tax"] = it }
        shipping?.let { map["shipping"] = it }
        discount?.let { map["discount"] = it }
        coupon?.let { map["coupon"] = it }
        currency?.let { map["currency"] = it }
        if (products.isNotEmpty()) {
            map["products"] = products.map { it.toMap() }
        }
        return map
    }

    fun toJSONObject(): JSONObject = JSONObject(toMap().filterValues { it != null })
}

// =============================================================================
// ECOMMERCE MODULE
// =============================================================================

object AetherEcommerce {
    private const val PREFS_NAME = "com.aether.sdk.ecommerce"
    private const val CART_KEY = "aether_cart"

    private var prefs: SharedPreferences? = null

    /** In-memory cart keyed by product ID. Thread-safe. */
    private val cart = ConcurrentHashMap<String, AetherCartItem>()

    /**
     * Initialize the ecommerce module. Call once from Application.onCreate().
     */
    fun initialize(context: Context) {
        prefs = context.applicationContext
            .getSharedPreferences(PREFS_NAME, Context.MODE_PRIVATE)
        loadCart()
    }

    // =========================================================================
    // PRODUCT EVENTS
    // =========================================================================

    /** Track when a user views a product. */
    fun productViewed(product: AetherProduct) {
        Aether.track("product_viewed", product.toMap())
    }

    /** Track when a user views a product list / collection. */
    fun productListViewed(listId: String, products: List<AetherProduct>) {
        val props = mutableMapOf<String, Any?>(
            "list_id" to listId,
            "products" to products.map { it.toMap() },
            "product_count" to products.size
        )
        Aether.track("product_list_viewed", props)
    }

    // =========================================================================
    // CART MANAGEMENT
    // =========================================================================

    /** Add a product to the cart (or update quantity if already present). */
    fun addToCart(product: AetherProduct, quantity: Int = 1) {
        synchronized(cart) {
            val existing = cart[product.id]
            if (existing != null) {
                cart[product.id] = AetherCartItem(product, existing.quantity + quantity)
            } else {
                cart[product.id] = AetherCartItem(product, maxOf(1, quantity))
            }
            persistCart()
        }

        val props = product.toMap().toMutableMap()
        props["quantity"] = quantity
        Aether.track("add_to_cart", props)
    }

    /** Remove a product from the cart by its ID. */
    fun removeFromCart(productId: String) {
        val removedProps: Map<String, Any?>
        synchronized(cart) {
            val item = cart.remove(productId)
            removedProps = item?.toMap() ?: mapOf("product_id" to productId)
            persistCart()
        }
        Aether.track("remove_from_cart", removedProps)
    }

    /** Track a cart-viewed event with all current items. */
    fun cartViewed() {
        Aether.track("cart_viewed", cartSnapshot())
    }

    /** Get all items currently in the cart. */
    fun getCart(): List<AetherCartItem> {
        return synchronized(cart) { cart.values.toList() }
    }

    /** Calculate total cart value (price * quantity for each item). */
    fun getCartValue(): Double {
        return synchronized(cart) {
            cart.values.sumOf { it.product.price * it.quantity }
        }
    }

    /** Remove all items from the cart. */
    fun clearCart() {
        synchronized(cart) {
            cart.clear()
            persistCart()
        }
    }

    // =========================================================================
    // CHECKOUT FUNNEL
    // =========================================================================

    /** Track the start of a checkout. */
    fun checkoutStarted(properties: Map<String, Any?> = emptyMap()) {
        val props = cartSnapshot().toMutableMap()
        props.putAll(properties)
        Aether.track("checkout_started", props)
    }

    /** Track individual checkout steps (e.g., shipping, payment, review). */
    fun checkoutStepCompleted(step: Int, properties: Map<String, Any?> = emptyMap()) {
        val props = mutableMapOf<String, Any?>("step" to step)
        props.putAll(properties)
        Aether.track("checkout_step_completed", props)
    }

    // =========================================================================
    // ORDER EVENTS
    // =========================================================================

    /** Track a completed purchase. */
    fun orderCompleted(order: AetherOrder) {
        Aether.track("order_completed", order.toMap())
        clearCart()
    }

    /** Track a full or partial refund. */
    fun orderRefunded(orderId: String, products: List<AetherProduct>? = null) {
        val props = mutableMapOf<String, Any?>("order_id" to orderId)
        if (!products.isNullOrEmpty()) {
            props["products"] = products.map { it.toMap() }
            props["full_refund"] = false
        } else {
            props["full_refund"] = true
        }
        Aether.track("order_refunded", props)
    }

    // =========================================================================
    // COUPONS
    // =========================================================================

    /** Track when a coupon is applied. */
    fun couponApplied(coupon: String, discount: Double? = null) {
        val props = mutableMapOf<String, Any?>("coupon" to coupon)
        discount?.let { props["discount"] = it }
        Aether.track("coupon_applied", props)
    }

    // =========================================================================
    // PERSISTENCE
    // =========================================================================

    private fun cartSnapshot(): Map<String, Any?> {
        return synchronized(cart) {
            mapOf(
                "products" to cart.values.map { it.toMap() },
                "cart_value" to cart.values.sumOf { it.product.price * it.quantity },
                "item_count" to cart.size
            )
        }
    }

    private fun persistCart() {
        val array = JSONArray()
        for (item in cart.values) {
            val obj = JSONObject().apply {
                put("product", item.product.toJSONObject())
                put("quantity", item.quantity)
            }
            array.put(obj)
        }
        prefs?.edit()?.putString(CART_KEY, array.toString())?.apply()
    }

    private fun loadCart() {
        val raw = prefs?.getString(CART_KEY, null) ?: return
        try {
            val array = JSONArray(raw)
            for (i in 0 until array.length()) {
                val obj = array.getJSONObject(i)
                val pObj = obj.getJSONObject("product")
                val product = AetherProduct(
                    id = pObj.getString("product_id"),
                    name = pObj.getString("name"),
                    category = pObj.optString("category", null),
                    brand = pObj.optString("brand", null),
                    variant = pObj.optString("variant", null),
                    price = pObj.optDouble("price", 0.0),
                    quantity = if (pObj.has("quantity")) pObj.getInt("quantity") else null,
                    currency = pObj.optString("currency", null),
                    coupon = pObj.optString("coupon", null)
                )
                val qty = obj.getInt("quantity")
                cart[product.id] = AetherCartItem(product, qty)
            }
        } catch (_: Exception) {
            // Corrupted cache — start fresh
            cart.clear()
        }
    }
}
